src/Microsoft.Azure.SignalR.Common/ClientInvocation/ClientResultsManager.cs

```diff
@@ -0,0 +1,160 @@
+    {
+        private readonly ConcurrentDictionary<string, PendingInvocation> _pendingInvocations = new();
+        private readonly ConcurrentDictionary<string, List<string>> _serviceMappingMessages = new();
+        private readonly string _serverGUID = Guid.NewGuid().ToString();
+        private long _lastInvocationId = 0;
+
+        private readonly IHubProtocolResolver _hubProtocolResolver;
+
+        public ClientResultsManager(IHubProtocolResolver hubProtocolResolver)
+        {
+            _hubProtocolResolver = hubProtocolResolver;
+        }
+
+        public string GetNewInvocationId(string connectionId)
+        {
+            return $"{connectionId}-{_serverGUID}-{Interlocked.Increment(ref _lastInvocationId)}";
+        }
+
+        public Task<T> AddInvocation<T>(string connectionId, string invocationId, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSourceWithCancellation<T>(
+                cancellationToken,
+                () => TryCompleteResult(connectionId, CompletionMessage.WithError(invocationId, "Canceled")));
+
+            var result = _pendingInvocations.TryAdd(invocationId, new PendingInvocation(typeof(T), connectionId, tcs, static (state, completionMessage) =>
+            {
+                var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+                if (completionMessage.HasResult)
+                {
+                    tcs.SetResult((T)completionMessage.Result);
+                }
+                else
+                {
+                    tcs.SetException(new Exception(completionMessage.Error));
+                }
+            }
+            ));
+            Debug.Assert(result);
+
+            tcs.RegisterCancellation();
+
+            return tcs.Task;
+        }
+
+        public void AddServiceMappingMessage(ServiceMappingMessage serviceMappingMessage)
+        {
+            _serviceMappingMessages.TryGetValue(serviceMappingMessage.InstanceId, out var oldValue);
+            var newValue = oldValue ?? new List<string> { };
+            newValue.Add(serviceMappingMessage.InvocationId);
+            _serviceMappingMessages.TryUpdate(serviceMappingMessage.InstanceId, newValue, oldValue);
+        }
+
+        public void CleanupInvocations(string instanceId)
+        {
+            foreach (var invocationId in _serviceMappingMessages[instanceId])
+            {
+                if (_pendingInvocations.TryRemove(invocationId, out var item))
+                {
+                    var message = new CompletionMessage(invocationId, $"Connection '{item.ConnectionId}' disconnected.", null, false);
+                    item.Complete(item.Tcs, message);
+                }
+            }
+        }
+
+        public bool TryCompleteResult(string connectionId, CompletionMessage message)
+        {
+            if (_pendingInvocations.TryGetValue(message.InvocationId!, out var item))
```

Comment:
directly call `TryRemove`?

Reply:
Direct call may lead to difference. The next lines wrote
```csharp
if (item.ConnectionId != connectionId)
{
    throw new InvalidOperationException($"Connection ID '{connectionId}' is not valid for invocation ID '{message.InvocationId}'.");
}
```
if `!=`, the corresponding invocation won't be removed. If a correct `connectionId` is passed in in the future, the invocation can still take effect